// jsdom/lib/jsdom/living/index.js
"use strict";

exports.DOMException = require("jsdom/lib/jsdom/web-idl/DOMException");
exports.NamedNodeMap = require("jsdom/lib/jsdom/living/attributes").NamedNodeMap;
exports.Attr = require("jsdom/lib/jsdom/living/generated/Attr").interface;
exports.Node = require("jsdom/lib/jsdom/living/generated/Node").interface;
exports.Element = require("jsdom/lib/jsdom/living/generated/Element").interface;
exports.DocumentFragment = require("jsdom/lib/jsdom/living/generated/DocumentFragment").interface;
exports.Document = exports.HTMLDocument = require("jsdom/lib/jsdom/living/generated/Document").interface;
exports.CharacterData = require("jsdom/lib/jsdom/living/generated/CharacterData").interface;
exports.Comment = require("jsdom/lib/jsdom/living/generated/Comment").interface;
exports.DocumentType = require("jsdom/lib/jsdom/living/generated/DocumentType").interface;
exports.DOMImplementation = require("jsdom/lib/jsdom/living/generated/DOMImplementation").interface;
exports.ProcessingInstruction = require("jsdom/lib/jsdom/living/generated/ProcessingInstruction").interface;
exports.Text = require("jsdom/lib/jsdom/living/generated/Text").interface;

exports.Event = require("jsdom/lib/jsdom/living/generated/Event").interface;
exports.CustomEvent = require("jsdom/lib/jsdom/living/generated/CustomEvent").interface;
exports.MessageEvent = require("jsdom/lib/jsdom/living/generated/MessageEvent").interface;
exports.ErrorEvent = require("jsdom/lib/jsdom/living/generated/ErrorEvent").interface;
exports.HashChangeEvent = require("jsdom/lib/jsdom/living/generated/HashChangeEvent").interface;
exports.PopStateEvent = require("jsdom/lib/jsdom/living/generated/PopStateEvent").interface;
exports.UIEvent = require("jsdom/lib/jsdom/living/generated/UIEvent").interface;
exports.MouseEvent = require("jsdom/lib/jsdom/living/generated/MouseEvent").interface;
exports.KeyboardEvent = require("jsdom/lib/jsdom/living/generated/KeyboardEvent").interface;
exports.TouchEvent = require("jsdom/lib/jsdom/living/generated/TouchEvent").interface;
exports.MutationEvent = require("jsdom/lib/jsdom/living/generated/MutationEvent").interface;
exports.ProgressEvent = require("jsdom/lib/jsdom/living/generated/ProgressEvent").interface;
exports.EventTarget = require("jsdom/lib/jsdom/living/generated/EventTarget").interface;

exports.Location = require("jsdom/lib/jsdom/living/generated/Location").interface;
exports.History = require("jsdom/lib/jsdom/living/generated/History").interface;

require("jsdom/lib/jsdom/living/register-elements")(exports);

// These need to be cleaned up...
require("jsdom/lib/jsdom/level2/style")(exports);
require("jsdom/lib/jsdom/level3/xpath")(exports);

// These are OK but need migration to webidl2js eventually.
require("jsdom/lib/jsdom/living/html-collection")(exports);
require("jsdom/lib/jsdom/living/node-filter")(exports);
require("jsdom/lib/jsdom/living/node-iterator")(exports);
require("jsdom/lib/jsdom/living/node-list")(exports);
require("jsdom/lib/jsdom/living/blob")(exports);
require("jsdom/lib/jsdom/living/file")(exports);
require("jsdom/lib/jsdom/living/filelist")(exports);
require("jsdom/lib/jsdom/living/form-data")(exports);
require("jsdom/lib/jsdom/living/xmlhttprequest-event-target")(exports);
require("jsdom/lib/jsdom/living/xmlhttprequest-upload")(exports);

exports.DOMTokenList = require("jsdom/lib/jsdom/living/dom-token-list").DOMTokenList;
exports.URL = require("whatwg-url").URL;
