// jsdom/lib/jsdom/living/register-elements.js
"use strict";
/* eslint global-require: 0 */

const DocumentImpl = require("jsdom/lib/jsdom/living/nodes/Document-impl.js");

const mappings = {
  HTMLElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLElement.js"),
    tags: []
  },
  HTMLAnchorElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLAnchorElement.js"),
    tags: ["a"]
  },
  HTMLAppletElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLAppletElement.js"),
    tags: ["applet"]
  },
  HTMLAreaElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLAreaElement.js"),
    tags: ["area"]
  },
  HTMLAudioElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLAudioElement.js"),
    tags: ["audio"]
  },
  HTMLBaseElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLBaseElement.js"),
    tags: ["base"]
  },
  HTMLBodyElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLBodyElement.js"),
    tags: ["body"]
  },
  HTMLBRElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLBRElement.js"),
    tags: ["br"]
  },
  HTMLButtonElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLButtonElement.js"),
    tags: ["button"]
  },
  HTMLCanvasElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLCanvasElement.js"),
    tags: ["canvas"]
  },
  HTMLDataElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLDataElement.js"),
    tags: ["data"]
  },
  HTMLDataListElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLDataListElement.js"),
    tags: ["datalist"]
  },
  HTMLDialogElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLDialogElement.js"),
    tags: ["dialog"]
  },
  HTMLDirectoryElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLDirectoryElement.js"),
    tags: ["dir"]
  },
  HTMLDivElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLDivElement.js"),
    tags: ["div"]
  },
  HTMLDListElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLDListElement.js"),
    tags: ["dl"]
  },
  HTMLEmbedElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLEmbedElement.js"),
    tags: ["embed"]
  },
  HTMLFieldSetElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLFieldSetElement.js"),
    tags: ["fieldset"]
  },
  HTMLFontElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLFontElement.js"),
    tags: ["font"]
  },
  HTMLFormElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLFormElement.js"),
    tags: ["form"]
  },
  HTMLFrameElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLFrameElement.js"),
    tags: ["frame"]
  },
  HTMLFrameSetElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLFrameSetElement.js"),
    tags: ["frameset"]
  },
  HTMLHeadingElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLHeadingElement.js"),
    tags: ["h1", "h2", "h3", "h4", "h5", "h6"]
  },
  HTMLHeadElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLHeadElement.js"),
    tags: ["head"]
  },
  HTMLHRElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLHRElement.js"),
    tags: ["hr"]
  },
  HTMLHtmlElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLHtmlElement.js"),
    tags: ["html"]
  },
  HTMLIFrameElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLIFrameElement.js"),
    tags: ["iframe"]
  },
  HTMLImageElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLImageElement.js"),
    tags: ["img"]
  },
  HTMLInputElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLInputElement.js"),
    tags: ["input"]
  },
  HTMLLabelElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLLabelElement.js"),
    tags: ["label"]
  },
  HTMLLegendElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLLegendElement.js"),
    tags: ["legend"]
  },
  HTMLLIElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLLIElement.js"),
    tags: ["li"]
  },
  HTMLLinkElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLLinkElement.js"),
    tags: ["link"]
  },
  HTMLMapElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLMapElement.js"),
    tags: ["map"]
  },
  HTMLMediaElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLMediaElement.js"),
    tags: []
  },
  HTMLMenuElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLMenuElement.js"),
    tags: ["menu"]
  },
  HTMLMetaElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLMetaElement.js"),
    tags: ["meta"]
  },
  HTMLMeterElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLMeterElement.js"),
    tags: ["meter"]
  },
  HTMLModElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLModElement.js"),
    tags: ["del", "ins"]
  },
  HTMLObjectElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLObjectElement.js"),
    tags: ["object"]
  },
  HTMLOListElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLOListElement.js"),
    tags: ["ol"]
  },
  HTMLOptGroupElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLOptGroupElement.js"),
    tags: ["optgroup"]
  },
  HTMLOptionElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLOptionElement.js"),
    tags: ["option"]
  },
  HTMLOutputElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLOutputElement.js"),
    tags: ["output"]
  },
  HTMLParagraphElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLParagraphElement.js"),
    tags: ["p"]
  },
  HTMLParamElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLParamElement.js"),
    tags: ["param"]
  },
  HTMLPreElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLPreElement.js"),
    tags: ["pre"]
  },
  HTMLProgressElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLProgressElement.js"),
    tags: ["progress"]
  },
  HTMLQuoteElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLQuoteElement.js"),
    tags: ["blockquote", "q"]
  },
  HTMLScriptElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLScriptElement.js"),
    tags: ["script"]
  },
  HTMLSelectElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLSelectElement.js"),
    tags: ["select"]
  },
  HTMLSourceElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLSourceElement.js"),
    tags: ["source"]
  },
  HTMLSpanElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLSpanElement.js"),
    tags: ["span"]
  },
  HTMLStyleElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLStyleElement.js"),
    tags: ["style"]
  },
  HTMLTableCaptionElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableCaptionElement.js"),
    tags: ["caption"]
  },
  HTMLTableCellElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableCellElement.js"),
    tags: []
  },
  HTMLTableColElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableColElement.js"),
    tags: ["col", "colgroup"]
  },
  HTMLTableDataCellElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableDataCellElement.js"),
    tags: ["td"]
  },
  HTMLTableElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableElement.js"),
    tags: ["table"]
  },
  HTMLTableHeaderCellElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableHeaderCellElement.js"),
    tags: ["th"]
  },
  HTMLTimeElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTimeElement.js"),
    tags: ["time"]
  },
  HTMLTitleElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTitleElement.js"),
    tags: ["title"]
  },
  HTMLTableRowElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableRowElement.js"),
    tags: ["tr"]
  },
  HTMLTableSectionElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTableSectionElement.js"),
    tags: ["thead", "tbody", "tfoot"]
  },
  HTMLTemplateElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTemplateElement.js"),
    tags: ["template"]
  },
  HTMLTextAreaElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTextAreaElement.js"),
    tags: ["textarea"]
  },
  HTMLTrackElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLTrackElement.js"),
    tags: ["track"]
  },
  HTMLUListElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLUListElement.js"),
    tags: ["ul"]
  },
  HTMLUnknownElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLUnknownElement.js"),
    tags: []
  },
  HTMLVideoElement: {
    file: require("jsdom/lib/jsdom/living/generated/HTMLVideoElement.js"),
    tags: ["video"]
  }
};

module.exports = core => {
  for (const interfaceName of Object.keys(mappings)) {
    const file = mappings[interfaceName].file;
    const tags = mappings[interfaceName].tags;

    core[interfaceName] = file.interface;

    for (const tagName of tags) {
      DocumentImpl.implementation.prototype._elementBuilders[tagName] = (document, elName) => {
        return file.create([], {
          core,
          ownerDocument: document,
          localName: elName || tagName.toUpperCase()
        });
      };
    }
  }
};
